nixos/modules/services/search/opensearch.nix

```diff
@@ -0,0 +1,221 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.opensearch;
+
+  opensearchConfig = ''
+    network.host: ${cfg.listenAddress}
+    cluster.name: ${cfg.cluster_name}
+    ${lib.optionalString cfg.single_node "discovery.type: single-node"}
+
+    http.port: ${toString cfg.port}
+    transport.port: ${toString cfg.tcp_port}
+
+    ${cfg.extraConf}
+  '';
+
+  configDir = cfg.dataDir + "/config";
+
+  opensearchYml = pkgs.writeTextFile {
+    name = "opensearch.yml";
+    text = opensearchConfig;
+  };
```

Comment:
Instead of build the config by hand, try `pkgs.formats.yaml` from RFC 42.

https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md